File: cwclock-mobile/src/screens/onboarding/OrgPickerScreen.js
import React, { useEffect } from "react";
import { FlatList, StyleSheet, Text, TouchableOpacity } from "react-native";
import { useDispatch, useSelector } from "react-redux";
import Screen from "../../components/Screen";
import { useI18n } from "../../i18n/I18nContext";
import { listOrganizationsApi } from "../../redux/organizations/organizations.actions";
import { selectOrgApi } from "../../redux/session/session.actions";
import theme from "../../theme";

// OrgPickerScreen shows whenever session.status is "needsOrg" (see
// session.reducer.js) - either onboarding didn't have an org_id line
// (manual entry / an older QR), or the user picked "Switch organization"
// from SettingsScreen. Selecting a row dispatches selectOrgApi and the root
// navigator (src/App.js) reacts to the resulting status change itself.
const OrgPickerScreen = ({ navigation }) => {
  const { t } = useI18n();
  const dispatch = useDispatch();
  const { items, isLoading } = useSelector((state) => state.organizations);

  useEffect(() => {
    dispatch(listOrganizationsApi());
  }, [dispatch]);

  const handleSelect = async (orgId) => {
    await dispatch(selectOrgApi(orgId));
    // No-op when there's nothing to go back to (first-run onboarding) -
    // only meaningful when reached from SettingsScreen's "Switch
    // organization", where session.status was already "connected".
    navigation.goBack();
  };

  return (
    <Screen scroll={false}>
      <Text style={styles.title}>{t("onboarding.pickOrganization")}</Text>
      <FlatList
        data={items}
        keyExtractor={(item) => item.id}
        contentContainerStyle={styles.list}
        ListEmptyComponent={!isLoading ? <Text style={styles.empty}>{t("onboarding.noOrganizations")}</Text> : null}
        renderItem={({ item }) => (
          <TouchableOpacity style={styles.row} onPress={() => handleSelect(item.id)}>
            <Text style={styles.rowText}>{item.name}</Text>
          </TouchableOpacity>
        )}
      />
    </Screen>
  );
};

const styles = StyleSheet.create({
  title: {
    fontSize: 20,
    fontWeight: "700",
    color: theme.color.text,
    padding: theme.spacing(2),
  },
  list: {
    paddingHorizontal: theme.spacing(2),
  },
  row: {
    paddingVertical: theme.spacing(2),
    borderBottomWidth: 1,
    borderBottomColor: theme.color.border,
  },
  rowText: {
    fontSize: 17,
    color: theme.color.text,
  },
  empty: {
    color: theme.color.textMuted,
    padding: theme.spacing(2),
  },
});

export default OrgPickerScreen;

File: cwclock-mobile/lib/screens/org_picker/org_picker_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../providers/locale_provider.dart';
import '../../providers/organizations_provider.dart';
import '../../providers/session_provider.dart';
import '../../theme.dart';

/// Shown whenever session status is "needsOrg" - either onboarding didn't
/// have an org_id line (manual entry / an older QR), or the user picked
/// "Switch organization" from Settings. Selecting a row calls
/// SessionNotifier.selectOrg and the router's redirect (see router.dart)
/// reacts to the resulting status change itself. Ported from
/// src/screens/onboarding/OrgPickerScreen.js.
class OrgPickerScreen extends ConsumerStatefulWidget {
  const OrgPickerScreen({super.key});

  @override
  ConsumerState<OrgPickerScreen> createState() => _OrgPickerScreenState();
}

class _OrgPickerScreenState extends ConsumerState<OrgPickerScreen> {
  @override
  void initState() {
    super.initState();
    Future.microtask(() => ref.read(organizationsProvider.notifier).listOrganizations());
  }

  Future<void> _handleSelect(String orgId) async {
    await ref.read(sessionProvider.notifier).selectOrg(orgId);
    // No-op when there's nothing to go back to (first-run onboarding) - only
    // meaningful when reached from Settings' "Switch organization", where
    // session status was already "connected".
    if (mounted && Navigator.of(context).canPop()) {
      Navigator.of(context).pop();
    }
  }

  @override
  Widget build(BuildContext context) {
    final locale = ref.watch(localeProvider);
    final t = translateWith(locale);
    final orgsState = ref.watch(organizationsProvider);

    return Scaffold(
      appBar: AppBar(title: Text(t('onboarding.pickOrganization'))),
      body: SafeArea(
        child: orgsState.items.isEmpty && !orgsState.isLoading
            ? Padding(
                padding: EdgeInsets.all(AppSpacing.of(2)),
                child: Text(t('onboarding.noOrganizations'), style: const TextStyle(color: AppColors.textMuted)),
              )
            : ListView.separated(
                padding: EdgeInsets.symmetric(horizontal: AppSpacing.of(2)),
                itemCount: orgsState.items.length,
                separatorBuilder: (_, _) => const Divider(height: 1, color: AppColors.border),
                itemBuilder: (context, index) {
                  final org = orgsState.items[index];
                  return ListTile(
                    contentPadding: EdgeInsets.zero,
                    title: Text(org.name, style: const TextStyle(fontSize: 17, color: AppColors.text)),
                    onTap: () => _handleSelect(org.id),
                  );
                },
              ),
      ),
    );
  }
}
